nav2_costmap_2d/plugins/denoise_layer.cpp

```diff
@@ -0,0 +1,244 @@
+  }
+
+  // wrap roi_image over existing costmap2d buffer
+  unsigned char * master_array = master_grid.getCharMap();
+  const int step = static_cast<int>(master_grid.getSizeInCellsX());
+
+  const size_t width = max_x - min_x;
+  const size_t height = max_y - min_y;
+  Image<uint8_t> roi_image(height, width, master_array + min_x, step);
+
+  try {
+    denoise(roi_image);
+  } catch (std::exception & ex) {
+    RCLCPP_ERROR(logger_, (std::string("Inner error: ") + ex.what()).c_str());
+  }
+
+  current_ = true;
+}
+
+void
+DenoiseLayer::denoise(Image<uint8_t> & image) const
+{
+  if (image.empty()) {
+    return;
+  }
+
+  if (minimal_group_size_ <= 1) {
+    return;  // A smaller group cannot exist. No one pixel will be changed
+  }
+
+  if (minimal_group_size_ == 2) {
+    // Performs fast filtration based on erosion function
+    removeSinglePixels(image);
+  } else {
+    // Performs a slower segmentation-based operation
+    removeGroups(image);
+  }
+}
+
+template<ConnectivityType connectivity, class Label>
+void removeGroupsImpl(Image<uint8_t> & image, MemoryBuffer & buffer, size_t minimal_group_size)
+{
+  // Creates an image labels in which each obstacles group is labeled with a unique code
+  auto components = ConnectedComponents<connectivity, Label>::detect(image, buffer, isBackground);
+  const Label groups_count = components.second;
+  const Image<Label> & labels = components.first;
+
+  // Calculates the size of each group.
+  // Group size is equal to the number of pixels with the same label
+  const Label max_label_value = groups_count - 1;  // It's safe. groups_count always non-zero
+  std::vector<size_t> groups_sizes = histogram(
+    labels, max_label_value, minimal_group_size + 1);
+
+  // The group of pixels labeled 0 corresponds to empty map cells.
+  // Zero bin of the histogram is equal to the number of pixels in this group.
+  // Because the values of empty map cells should not be changed, we will reset this bin
+  groups_sizes.front() = 0;  // don't change image background value
+
+  // noise_labels_table[i] = true if group with label i is noise
+  std::vector<bool> noise_labels_table(groups_sizes.size());
+  auto transform_fn = [&minimal_group_size](size_t bin_value) {
+      return bin_value < minimal_group_size;
+    };
+  std::transform(
+    groups_sizes.begin(), groups_sizes.end(), noise_labels_table.begin(),
+    transform_fn);
+
+  // Replace the pixel values from the small groups to background code
+  labels.convert(
+    image, [&noise_labels_table](Label src, uint8_t & trg) {
+      if (!isBackground(trg) && noise_labels_table[src]) {
+        trg = 0;
+      }
+    });
+}
+
+template<ConnectivityType connectivity>
+void removeGroupsPickLabelType(
+  Image<uint8_t> & image, MemoryBuffer & buffer,
+  size_t minimal_group_size)
+{
+  if (ConnectedComponents<connectivity, uint32_t>::optimalBufferSize(image) <= buffer.capacity()) {
+    removeGroupsImpl<connectivity, uint32_t>(image, buffer, minimal_group_size);
```

Comment:
Can you explain the reason for the templating on the unsigned int sizes?

Reply:
See [this](https://github.com/ros-planning/navigation2/pull/2567#discussion_r758451829) comment. The second template parameter of the `removeGroupsImpl` - the type of label.